[DNM]: sql: streaming internal executor sanity check

yuzefovich opened the pull request.

yuzefovich committed:
sql: make internal executor streaming

This commit updates the internal executor to operate in a streaming
fashion by refactoring its internal logic to implement an iterator
pattern. A new method `QueryInternalEx` (and its counterpart
`QueryInternal`) is introduced (both not used currently) while all
existing methods of `InternalExecutor` interface are implemented
using the new iterator logic.

The communication between the iterator goroutine (the receiver) and the
connExecutor goroutine (the sender) is done via a buffered (of 32 size
in non-test setting) channel. The channel is closed by one of the two
callbacks given to the connExecutor. Additionally, the sync callback is
responsible for sending the 'rows affected' row as the last one on the
channel (if it is required).

Care needs to be taken when closing the iterator - we need to make sure
to close the stmtBuf (so that there are no more commands for the
connExecutor goroutine to execute) and then we need to unblockingly
drain the channel (since the connExecutor goroutine might be blocked on
adding a row to the channel). After that we have to wait for the
connExecutor goroutine to exit so that we can finish the tracing span.
For convenience purposes, if the iterator is fully exhausted, it will
get closed automatically.

Release note: None

yuzefovich committed:
sqlutil: remove Query and QueryEx from InternalExecutor interface

All usages of `sqlutil.InternalExecutor.Query` and
`sqlutil.InternalExecutor.QueryEx` have been refactored: most of
the changes are taking advantage of the iterator API, in a couple
of places we now use `Exec` and `ExecEx` methods since in those
places we only need the number of affected rows.

In a few places where we still need to buffer all rows, the caller
currently performs the buffering. In the follow-up commit
`QueryBuffered` and `QueryBufferedEx` will be added into the interface
(essentially renaming `Query` and `QueryEx` to make it explicit that the
buffering occurs), and the custom logic will be removed.

Release note: None

yuzefovich committed:
sqlutil: remove QueryWithCols in favor of QueryRowExWithCols

This commit refactors all usages of
`sqlutil.InternalExecutor.QueryWithCols` method in favor of newly added
`QueryRowExWithCols` since in almost all places where the former was
used, we expected exactly one row. The only noticeable change is the
refactor of `jobScheduler.executeSchedules` where we now use the
iterator pattern. The difference there is that now it is possible to
execute some schedules before encountering an error on the iterator
(previously, we would buffer up all rows first), but this change seems
reasonable to me.

Release note: None

yuzefovich committed:
sql: remove QueryWithCols method from the internal executor

Previous commit removed this method from the interface, and this commit
follows up to remove the method entirely. This is done in a similar
fashion - by changing to using `QueryRowExWithCols` (when at most one
row is expected) and to using the iterator API (avoiding the buffering
of rows in all cases).

Release note: None

yuzefovich committed:
sql: audit all usages of QueryEx to use iterator pattern

This commit audits the usage of `QueryEx` method of the internal
executor in the following manner:
- if the caller only needed to execute the statement, `ExecEx` is now
used
- if the query can return at most one row, then `QueryRowEx` is now used
- if the caller can be refactored to use the iterator pattern, it is
done so.

As a result, almost all usages have been refactored (most notably the
virtual `crdb_internal.jobs` page now uses the iterator pattern, thus
aleviating OOM concerns - the ad-hoc memory accounting logic has been
removed).

`QueryEx` has been renamed to `QueryBufferedEx` to highlight that the
full buffering occurs, and it was added to `sqlutil.InternalExecutor`
interface. The method is now used only in two places.

Release note: None

yuzefovich committed:
sql: audit all usages of Query to use iterator pattern

Similarly to the previous commit, here we audit all usages of `Query`
method of the internal executor to take advantage of the iterator API
wherever possible (or switching to `Exec` or `QueryRow`).
`QueryBuffered` has been added to the interface too.

The only place where it would be beneficial to use the iterator pattern
but it is not done currently is for `SHOW STATISTICS` statement - in
there, we have a panic-catcher which works only on the assumption of not
updating any of the shared state (which the iterator API contradicts).
Refactoring that part is left as a TODO.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59688)
<!-- Reviewable:end -->

yuzefovich committed:
sql: remove resWithPos from the internal executor

This is a WIP commit that will be squashed with the previous one if we
decide to keep it. It is an attempt to remove the accumulation of
results for each of the commands put onto the stmtBuf of the internal
executor. This commit removes the ability to auto-retry in the executor
itself.

Release note: None
